### ISSUE TITLE
keep_by_idx() returns list of empty elements.

### ISSUE THREAD

**Issue by lizhuimu** [Dobiasd/FunctionalPlus#243]
`vector<string> const strings{"str1", "str2"};`                                                             
`auto result = fplus::keep_by_idx([](auto idx) { return true; }, strings);`       

The result should be {"str1", "str2"}, but it is actually {"", ""}.
I think line 114 in filter.hpp has something to do with the problem.

**Comment by jee-lee** [comment]
Turns out if UnaryPredicate returns true for idx 0, all the elements keep_by_idx() returns will be empty string.

**Comment by Dobiasd** [comment]
Oh, thanks for reporting! This seems like a severe bug. I'm afk-ish for the rest of this month, but I'll look into this as soon as I'm back.

**Comment by Dobiasd** [comment]
> I think line 114 in filter.hpp has something to do with the problem.

You're absolutely right.

The current test case uses `int`. Self-assigning (which happens when the first element of the vector is kept) is fine:

```cpp
int one = 1;
one = std::move(one);
assert(one == 1); // ok.
```

But not with `std::string` (and probably a lot of other types):

```cpp
std::string foo = "foo";
foo = std::move(foo);
assert(foo == "foo"); // fail!
```

I'll implement a fix and also check if this mistake was done at other places too. Thanks again! :slightly_smiling_face:

**Comment by Dobiasd** [comment]
I've just re-opened this issue, because the pull request, that will fix it, is not yet merged. Once merged, the `fixes #243` part in the commit message will auto-close this issue. :slightly_smiling_face:
